unit-tests/run-unit-tests.py

```diff
@@ -264,79 +267,111 @@ def check_log_for_fails(log, testname, exe):
             return True
         return False
 
-logdir = target + '/unit-tests'
-os.makedirs( logdir, exist_ok = True )
-n_tests = 0
-
-# In Linux, the build targets are located elsewhere than on Windows
-# Go over all the tests from a "manifest" we take from the result of the last CMake
-# run (rather than, for example, looking for test-* in the build-directory):
-if linux:
-    manifestfile = target + '/CMakeFiles/TargetDirectories.txt'
-else:
-    manifestfile = target + '/../CMakeFiles/TargetDirectories.txt'
-
-for manifest_ctx in grep( r'(?<=unit-tests/build/)\S+(?=/CMakeFiles/test-\S+.dir$)', manifestfile ):
-    
-    testdir = manifest_ctx['match'].group(0)                    # "log/internal/test-all"
-    testparent = os.path.dirname(testdir)                       # "log/internal"
-    if testparent:
-        testname = 'test-' + testparent.replace( '/', '-' ) + '-' + os.path.basename(testdir)[5:]    # "test-log-internal-all"
-    else:
-        testname = testdir                                      # no parent folder so we get "test-all"
-    if linux:
-        exe = target + '/unit-tests/build/' + testdir + '/' + testname
-    else:
-        exe = target + '/' + testname + '.exe'
-    log = logdir + '/' + testname + '.log'
+# definition of classes for tests
+class Test(ABC):
+    """
+    Abstract class for a test. Holds the name of the test, the log file for it and the command line needed to run it
+    """
+    def __init__(self, testname, command, executable):
+        self.testname = testname
+        self.command = command
+        self.executable = executable
+        self.log = logdir + '/' + testname + '.log'
 
-    progress( testname, '>', log, '...' )
-    n_tests += 1
-    try:
-        run( [exe], stdout=log )
-    except FileNotFoundError:
-        error( red + testname + reset + ': executable not found! (' + exe + ')' )
-        continue
-    except subprocess.CalledProcessError as cpe:
-        if not check_log_for_fails(log, testname, exe):
-            # An unexpected error occurred
-            error( red + testname + reset + ': exited with non-zero value! (' + str(cpe.returncode) + ')' )
-
-# If we run python tests with no .pyd/.so file they will crash. Therefore we only run them if such a file was found
-if pyrs:
-    # unit-test scripts are in the same directory as this script
-    for py_test in find(current_dir, '(^|/)test-.*\.py'):
-    
-        testdir = py_test[:-3]                         # "log/internal/test-all"  <-  "log/internal/test-all.py"
-        testparent = os.path.dirname(testdir)          # same as for cpp files
-        if testparent:
-            testname = 'test-' + testparent.replace( '/', '-' ) + '-' + os.path.basename(testdir)[5:]
-        else:
-            testname = testdir
-    
-        log = logdir + '/' + testname + '.log'
-        
-        progress( testname, '>', log, '...' )
+    def run_test(self):
```

Comment:
Why not just `run()`?

Reply:
And then... why not an abstract function that's implemented differently for Python vs. an exe?

Reply:
didn't want it to clash with the run function (the global one). Why abstract if it is the same for all tests?

Reply:
But you're using the base class to dictate the running command of the derived classes by enforcing the command from the ctor from the derived class. So why do it at all? Why not determine it at runtime, without using data?